clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -796,6 +796,126 @@ test_synthetic()
+    sched_ops.flags = static_cast<scheduler_t::scheduler_flags_t>(
+        static_cast<int>(sched_ops.flags) |
+        static_cast<int>(scheduler_t::SCHEDULER_SPECULATE_NOPS));
+    if (scheduler.init(sched_inputs, 1, sched_ops) != scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    int ordinal = 0;
+    auto *stream = scheduler.get_stream(0);
+    memref_t memref;
+    for (scheduler_t::stream_status_t status = stream->next_record(memref);
+         status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+        assert(status == scheduler_t::STATUS_OK);
+        switch (ordinal) {
+        case 0:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_PAGE_SIZE);
+            break;
+        case 1:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP);
+            break;
+        case 2:
+            assert(memref.marker.type == TRACE_TYPE_MARKER);
+            assert(memref.marker.marker_type == TRACE_MARKER_TYPE_CPU_ID);
+            break;
+        case 3:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 1);
+            break;
+        case 4:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 2);
+            // We realize now that we mispredicted that the branch would be taken.
+            // We ask to queue this record for post-speculation.
+            stream->start_speculation(100, true);
+            break;
+        case 5:
+            // We should now see nops from the speculator.
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 100);
+            break;
+        case 6:
+            // Another nop before we abandon this path.
+            assert(type_is_instr(memref.instr.type));
+#ifdef AARCH64
+            assert(memref.instr.addr == 104);
+#elif defined(X86_64) || defined(X86_32)
+            assert(memref.instr.addr == 101);
+#elif defined(ARM)
+            assert(memref.instr.addr == 102 || memref.instr.addr == 104);
+#endif
+            stream->stop_speculation();
+            break;
+        case 7:
+            // Back to the trace, to the queued record
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 2);
+            break;
+        case 8:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 3);
+            break;
+        case 9:
+            assert(type_is_instr(memref.instr.type));
+            assert(memref.instr.addr == 4);
+            // We realize now that we mispredicted that the branch would be taken.
+            // This time we do *not* ask to queue this record for post-speculation.
+            stream->start_speculation(200, false);
```

Comment:
Maybe also test multiple start_speculation calls to confirm that PC is changed after each?

Reply:
Done.